# Create a Logistic Regression to predict absenteeism

## import the relevant lib

In [1]:
import pandas as pd
import numpy as np

## Load data

In [2]:
data_preprocessed=pd.read_csv('Absenteeism_preprocessed.csv')

In [3]:
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

## Create the target

- as we are using Logistic regression its the problem of classification
- we need to classify the object into two category 'Moderatly absent' and 'Excessively absent'
- we will take the median value of 'Absenteeisum Time in Hours' and use it as a cut-off line that every thing
  below the median will be consider as normal and every thing above as excessive


In [4]:
data_preprocessed['Absenteeism Time in Hours'].median()

3.0

- Moderately absent (<= 3hours)  = 0
- Excessively absent (>=3 hours) = 1

In [5]:
targets=np.where(data_preprocessed['Absenteeism Time in Hours']>
                 data_preprocessed['Absenteeism Time in Hours'].median(),1,0)

In [6]:
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [7]:
data_preprocessed['Excessively Absenteeism']=targets

In [8]:
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0               30          0         2     1                          4   
1               31          0         1     0                          0   
2               31          0         0     0                          2   
3               24          0         2     0                          4   
4               30          0         2     1                          2   

   Excessively Absenteeism  
0                        1  
1                        0  
2                        0  
3                        1  
4                        0

In [9]:
(targets.sum()/targets.shape[0])*100

45.57142857142858

In [10]:
#checkpoint
data_with_targets=data_preprocessed.drop(['Absenteeism Time in Hours','Day of the Week',
                                          'Daily Work Load Average','Distance to Work'],axis=1)

## Input for regression

In [11]:
data_with_targets.shape

(700, 12)

In [12]:
unscaled_input=data_with_targets.iloc[:,:-1]

## Standardize the data

In [13]:
#from sklearn.preprocessing import StandardScaler

#absenteeism_scalar=StandardScaler()

In [14]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler

class CustomScaler(BaseEstimator,TransformerMixin): 
    def __init__(self,columns,copy=True,with_mean=True,with_std=True):
        self.scaler = StandardScaler(copy,with_mean,with_std)
        self.columns = columns
        self.mean_ = None
        self.var_ = None
    
    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns])
        self.var_ = np.var(X[self.columns])
        return self

    def transform(self, X, y=None, copy=None):
        init_col_order = X.columns
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]

In [15]:
unscaled_input.columns.values

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Transportation Expense', 'Age', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [16]:
#columns_to_scale= ['Month Value', 'Transportation Expense', 'Age','Body Mass Index', 'Education','Children', 'Pets']

columns_to_omit=['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4','Education']
columns_to_scale=[x for x in unscaled_input.columns.values if x not in columns_to_omit ]

In [17]:
absenteeism_scaler = CustomScaler(columns_to_scale)

C:\Users\yeeme\anaconda3\lib\site-packages\sklearn\utils\validation.py:70: FutureWarning: Pass copy=True, with_mean=True, with_std=True as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error
  warnings.warn(f"Pass {args_msg} as keyword args. From version "


In [18]:
absenteeism_scaler.fit(unscaled_input)

AttributeError: 'CustomScaler' object has no attribute 'copy'

AttributeError: 'CustomScaler' object has no attribute 'copy'

In [19]:
scaled_inputs = absenteeism_scaler.transform(unscaled_input)

In [20]:
scaled_inputs

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  \
0           0         0         0         1     0.182726   
1           0         0         0         0     0.182726   
2           0         0         0         1     0.182726   
3           1         0         0         0     0.182726   
4           0         0         0         1     0.182726   
..        ...       ...       ...       ...          ...   
695         1         0         0         0    -0.388293   
696         1         0         0         0    -0.388293   
697         1         0         0         0    -0.388293   
698         0         0         0         1    -0.388293   
699         0         0         0         1    -0.388293   

     Transportation Expense       Age  Body Mass Index  Education  Children  \
0                  1.005844 -0.536062         0.767431          0  0.880469   
1                 -1.574681  2.130803         1.002633          0 -0.019280   
2                 -0.654143  0.248310         1.002633          0 -0.919030   
3                  0.854936  0.405184        -0.643782          0  0.880469   
4                  1.005844 -0.536062         0.767431          0  0.880469   
..                      ...       ...              ...        ...       ...   
695               -0.654143  0.562059        -1.114186          1  0.880469   
696                0.040034 -1.320435        -0.643782          0 -0.019280   
697                1.624567 -1.320435        -0.408580          1 -0.919030   
698                0.190942 -0.692937        -0.408580          1 -0.919030   
699                1.036026  0.562059        -0.408580          0 -0.019280   

         Pets  
0    0.268487  
1   -0.589690  
2   -0.589690  
3   -0.589690  
4    0.268487  
..        ...  
695 -0.589690  
696  1.126663  
697 -0.589690  
698 -0.589690  
699  0.268487  

[700 rows x 11 columns]

In [21]:
scaled_inputs.shape

(700, 11)

## Split the data into Train & Test and Shuffle

## Import the relevant Module

In [22]:
from sklearn.model_selection import train_test_split

## Split

In [23]:
X_train,X_test,y_train,y_test=train_test_split(scaled_inputs,targets,test_size=0.2,random_state=20)

In [24]:
print(X_train.shape,y_train.shape)

(560, 11) (560,)


In [25]:
print(X_test.shape,y_test.shape)

(140, 11) (140,)


# TRain the MOdel

## Import relevant Module

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [27]:
log_reg=LogisticRegression()
log_reg.fit(X_train,y_train)

LogisticRegression()

In [28]:
log_reg.score(X_train,y_train)

0.7732142857142857

## Manually Check the accuracy

In [29]:
model_output=log_reg.predict(X_train)

In [30]:
model_output

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,

In [31]:
y_train

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,

In [32]:
np.sum((model_output == y_train))/model_output.shape[0]

0.7732142857142857

- <b>Based on the Data we used ,our model learned to classify ~80%(0.7732) of the observation correctly

## The intercept and coefficients

In [33]:
log_reg.intercept_

array([-1.6474549])

In [34]:
log_reg.coef_

array([[ 2.80019733,  0.95188356,  3.11555338,  0.83900082,  0.1589299 ,
         0.60528415, -0.16989096,  0.27981088, -0.21053312,  0.34826214,
        -0.27739602]])

In [35]:
feature_name=unscaled_input.columns.values

In [36]:
summary_table=pd.DataFrame(columns=['Feature_name'],data=feature_name)

summary_table['Coefficient']=np.transpose(log_reg.coef_)

summary_table

Feature_name  Coefficient
0                 Reason_1     2.800197
1                 Reason_2     0.951884
2                 Reason_3     3.115553
3                 Reason_4     0.839001
4              Month Value     0.158930
5   Transportation Expense     0.605284
6                      Age    -0.169891
7          Body Mass Index     0.279811
8                Education    -0.210533
9                 Children     0.348262
10                    Pets    -0.277396

In [37]:
summary_table.index=summary_table.index+1
summary_table.loc[0]=['Intercept',log_reg.intercept_[0]]

In [38]:
summary_table=summary_table.sort_index()
summary_table

Feature_name  Coefficient
0                Intercept    -1.647455
1                 Reason_1     2.800197
2                 Reason_2     0.951884
3                 Reason_3     3.115553
4                 Reason_4     0.839001
5              Month Value     0.158930
6   Transportation Expense     0.605284
7                      Age    -0.169891
8          Body Mass Index     0.279811
9                Education    -0.210533
10                Children     0.348262
11                    Pets    -0.277396

## Interpreting the Coefficients

In [39]:
summary_table['odds_ratio']=np.exp(summary_table.Coefficient)
summary_table

Feature_name  Coefficient  odds_ratio
0                Intercept    -1.647455    0.192539
1                 Reason_1     2.800197   16.447892
2                 Reason_2     0.951884    2.590585
3                 Reason_3     3.115553   22.545903
4                 Reason_4     0.839001    2.314054
5              Month Value     0.158930    1.172256
6   Transportation Expense     0.605284    1.831773
7                      Age    -0.169891    0.843757
8          Body Mass Index     0.279811    1.322880
9                Education    -0.210533    0.810152
10                Children     0.348262    1.416604
11                    Pets    -0.277396    0.757754

In [40]:
summary_table.sort_values('odds_ratio',ascending=False)

Feature_name  Coefficient  odds_ratio
3                 Reason_3     3.115553   22.545903
1                 Reason_1     2.800197   16.447892
2                 Reason_2     0.951884    2.590585
4                 Reason_4     0.839001    2.314054
6   Transportation Expense     0.605284    1.831773
10                Children     0.348262    1.416604
8          Body Mass Index     0.279811    1.322880
5              Month Value     0.158930    1.172256
7                      Age    -0.169891    0.843757
9                Education    -0.210533    0.810152
11                    Pets    -0.277396    0.757754
0                Intercept    -1.647455    0.192539

 A feature is not particularly important:
   - if its coefficient is around 0
   - if its coefficient is around 1

## Testing the model

In [43]:
log_reg.score(X_test,y_test)

0.75

In [44]:
predicted_proba=log_reg.predict_proba(X_test)
predicted_proba

array([[0.71340413, 0.28659587],
       [0.58724228, 0.41275772],
       [0.44020821, 0.55979179],
       [0.78159464, 0.21840536],
       [0.08410854, 0.91589146],
       [0.33487603, 0.66512397],
       [0.29984576, 0.70015424],
       [0.13103971, 0.86896029],
       [0.78625404, 0.21374596],
       [0.74903632, 0.25096368],
       [0.49397598, 0.50602402],
       [0.22484913, 0.77515087],
       [0.07129151, 0.92870849],
       [0.73178133, 0.26821867],
       [0.30934135, 0.69065865],
       [0.5471671 , 0.4528329 ],
       [0.55052275, 0.44947725],
       [0.5392707 , 0.4607293 ],
       [0.40201117, 0.59798883],
       [0.05361575, 0.94638425],
       [0.7003009 , 0.2996991 ],
       [0.78159464, 0.21840536],
       [0.42037128, 0.57962872],
       [0.42037128, 0.57962872],
       [0.24783565, 0.75216435],
       [0.74566259, 0.25433741],
       [0.51017274, 0.48982726],
       [0.85690195, 0.14309805],
       [0.20349733, 0.79650267],
       [0.78159464, 0.21840536],
       [0.

In [45]:
predicted_proba.shape

(140, 2)

In [46]:
predicted_proba[:,1]

array([0.28659587, 0.41275772, 0.55979179, 0.21840536, 0.91589146,
       0.66512397, 0.70015424, 0.86896029, 0.21374596, 0.25096368,
       0.50602402, 0.77515087, 0.92870849, 0.26821867, 0.69065865,
       0.4528329 , 0.44947725, 0.4607293 , 0.59798883, 0.94638425,
       0.2996991 , 0.21840536, 0.57962872, 0.57962872, 0.75216435,
       0.25433741, 0.48982726, 0.14309805, 0.79650267, 0.21840536,
       0.36956558, 0.67906035, 0.68502567, 0.52868083, 0.21840536,
       0.53506551, 0.22147081, 0.73692105, 0.40498044, 0.60505988,
       0.21075848, 0.45224466, 0.23751292, 0.39833498, 0.82755447,
       0.56797575, 0.69113325, 0.28659587, 0.21935267, 0.2033097 ,
       0.57628256, 0.3294664 , 0.66512397, 0.26949499, 0.83321968,
       0.43491525, 0.88374612, 0.23127072, 0.33415858, 0.34432939,
       0.69909345, 0.65494263, 0.29244941, 0.79200758, 0.20750276,
       0.26840558, 0.08708566, 0.22147081, 0.73245417, 0.30530219,
       0.22147081, 0.29014408, 0.90438191, 0.46061297, 0.60174

# Saving the Model

In [47]:
import pickle

In [48]:
with open('model','wb') as file:
    pickle.dump(log_reg,file)

In [49]:
with open('scaler','wb') as file:
    pickle.dump(absenteeism_scaler,file)